### PR TITLE
Moved party mode, set default and clear default to new contextmenu

### PR DESCRIPTION
**Pull request by razzeee**
This is a weird one.

Everything seems to work, but not really. The new actions show up fine in the places where they should be, but:
- Partymode doesn't seem to do anything for me (same with master)
- Set/Clear default doesn't seem to be supported by estuary/confluence (yes I know we talked about this)

But well moved stuff to the new structure.

### CODE REVIEW

**Comment by Razzeee** [comment]
Rebased. Looked into partymode and figured out that it's still working fine.

**Comment by Razzeee** [comment]
jenkins build and merge

**Comment by Razzeee** [comment]
jenkins build and merge

**Comment by MartijnKaijser** [comment]
PR has a merge commit

**Comment by Razzeee** [comment]
jenkins build and merge

File: xbmc/music/ContextMenus.cpp [hunk 1]
```diff
@@ -22,6 +22,14 @@
 #include "guilib/GUIWindowManager.h"
 #include "music/dialogs/GUIDialogMusicInfo.h"
 #include "tags/MusicInfoTag.h"
+#include "PartyModeManager.h"
+#include "settings/Settings.h"
+#include "utils/win32/Win32Log.h"
```
Comment (Rechi) [inline xbmc/music/ContextMenus.cpp]: you can't include this header on non windows platforms

Comment (Razzeee) [inline xbmc/music/ContextMenus.cpp]: Makes sense, not even sure where I got that from (Resharper I'm looking at you)

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by tamland** [comment]
@Razzeee This 'make default' context menu is visible for all kinds of things that make no sense now. Eg favourites, addons and all top level video nodes.

**Comment by Razzeee** [comment]
@tamland 
will investigate, as it should have to be one of these three nodes to show, if I'm reading the code right https://github.com/xbmc/xbmc/pull/10892/files#diff-79b7eb238e6484b2aa46c9739f99e168R73

There has to be something else going wrong

**Comment by ksooo** [comment]
I also just stumbled about this. :-/

**Comment by ksooo** [comment]
It seems, a url protocol check is missing somewhere, maybe in the new context menu item's visibility checking code. If you pass 'non-music' urls to CMusicDatabaseDirectory::GetDirectoryType, it will return <code>XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_ROOT<code>.

**Comment by ksooo** [comment]
Must the protocol of the matching item's path always be "musicdb"?

**Comment by ksooo** [comment]
Something like this?

<pre>
diff --git a/xbmc/music/ContextMenus.cpp b/xbmc/music/ContextMenus.cpp
index 593ac9c35a..a498c41a7d 100644
--- a/xbmc/music/ContextMenus.cpp
+++ b/xbmc/music/ContextMenus.cpp
@@ -25,6 +25,7 @@
 #include "PartyModeManager.h"
 #include "settings/Settings.h"
 #include "utils/LegacyPathTranslation.h"
+#include "utils/URIUtils.h"
 #include "utils/log.h"
 #include "filesystem/MusicDatabaseDirectory/DirectoryNode.h"
 #include "Util.h"
@@ -35,6 +36,26 @@
 namespace CONTEXTMENU
 {
 
+static bool IsMusicDbItem(const CFileItem& item)
+{
+  if (URIUtils::IsProtocol(item.GetPath(), "--kolja-put-the-right-protocol-in-here--"))
+  {
+    // are we in the playlists location?
+    bool inPlaylists = item.IsPath(CUtil::MusicPlaylistsLocation()) || item.IsPath("special://musicplaylists/");
+    //Set default or clear default
+    XFILE::CMusicDatabaseDirectory dir;
+    XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE nodetype = dir.GetDirectoryType(item.GetPath());
+    if (!item.IsParentFolder() && !inPlaylists &&
+        (nodetype == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_ROOT ||
+         nodetype == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_OVERVIEW ||
+         nodetype == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_TOP100))
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
 CMusicInfo::CMusicInfo(MediaType mediaType)
       : CStaticContextMenuAction(19033), m_mediaType(mediaType) {}
 
@@ -62,21 +83,9 @@ bool CPlayPartymode::Execute(const CFileItemPtr& item) const
 
 bool CSetDefault::IsVisible(const CFileItem& item) const
 {
-  if (!item.IsPath("sources://music/"))
-  {
-    // are we in the playlists location?
-    bool inPlaylists = item.IsPath(CUtil::MusicPlaylistsLocation()) || item.IsPath("special://musicplaylists/");
-    //Set default or clear default
-    XFILE::CMusicDatabaseDirectory dir;
-    XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE nodetype = dir.GetDirectoryType(item.GetPath());
-    if (!item.IsParentFolder() && !inPlaylists &&
-      (nodetype == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_ROOT ||
-        nodetype == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_OVERVIEW ||
-        nodetype == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_TOP100))
-    {
-      return !item.IsPath(CServiceBroker::GetSettings().GetString(CSettings::SETTING_MYMUSIC_DEFAULTLIBVIEW));
-    }
-  }
+  if (IsMusicDbItem(item))
+    return !item.IsPath(CServiceBroker::GetSettings().GetString(CSettings::SETTING_MYMUSIC_DEFAULTLIBVIEW));
+
   return false;
 }
 
@@ -126,21 +135,9 @@ bool CSetDefault::Execute(const CFileItemPtr& item) const
 
 bool CClearDefault::IsVisible(const CFileItem& item) const
 {
-  if (!item.IsPath("sources://music/"))
-  {
-    // are we in the playlists location?
-    bool inPlaylists = item.IsPath(CUtil::MusicPlaylistsLocation()) || item.IsPath("special://musicplaylists/");
-    //Set default or clear default
-    XFILE::CMusicDatabaseDirectory dir;
-    XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE nodetype = dir.GetDirectoryType(item.GetPath());
-    if (!item.IsParentFolder() && !inPlaylists &&
-      (nodetype == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_ROOT ||
-        nodetype == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_OVERVIEW ||
-        nodetype == XFILE::MUSICDATABASEDIRECTORY::NODE_TYPE_TOP100))
-    {
-      return !CServiceBroker::GetSettings().GetString(CSettings::SETTING_MYMUSIC_DEFAULTLIBVIEW).empty();
-    }
-  }
+  if (IsMusicDbItem(item))
+    return !CServiceBroker::GetSettings().GetString(CSettings::SETTING_MYMUSIC_DEFAULTLIBVIEW).empty();
+
   return false;
 }
 
</pre>

Side note: I would have bugged you with this PR before merge if I would have been aware of this PR, because of the 99.5% code copy for the IsVisible implementations. Shame on you. ;-)

**Comment by ksooo** [comment]
> Must the protocol of the matching item's path always be "musicdb"?
@Razzeee: Assuming my assumption is true (I have no idea how to trigger the functionality in question), would you think this code would fix things? => https://github.com/ksooo/xbmc/commit/77b37c46a4d7dd4c9612cf4d5db5264a92eff596

**Comment by Razzeee** [comment]
Promissing, but doesn't it miss `!item.IsParentFolder() && !inPlaylists`? Is that by design?
It might not be needed, as NODE_TYPE_ROOT, NODE_TYPE_OVERVIEW, NODE_TYPE_TOP100 might never be a in a playlist or not a parent folder.

**Comment by ksooo** [comment]
> Promissing, but doesn't it miss !item.IsParentFolder() && !inPlaylists? Is that by design?

 No and yes.

 As we now first check the url of the item against protocol "musicdb", there is neither need for an additional check for parent folder nor playlist as both do not have "musicdb" urls.

**Comment by Razzeee** [comment]
Got it, should I do a local test build or have you already done some testing in the wild?

**Comment by ksooo** [comment]
Please runtime test this. As I said, I have no idea how to actually test this. I have no plan where in Kodi I can find the nodes in question. Please also look that the context menu items do not appear where they shouldn't (example: context menu of addons never should have those items).

**Pull request merged**